framework/docs/source/tutorial-series-use-a-federated-learning-strategy-pytorch.rst

```diff
@@ -0,0 +1,421 @@
+
+.. code-block:: shell
+
+    $ flwr run .
+
+As we can see, the client logs now include the current round of federated learning
+(which they read from the ``config`` dictionary). We can also configure local training
+to run for one epoch during the first and second round of federated learning, and then
+for two epochs during the third round.
+
+Clients can also return arbitrary values to the server. To do so, they return a
+dictionary from ``fit`` and/or ``evaluate``. We have seen and used this concept
+throughout this tutorial without mentioning it explicitly: our ``FlowerClient`` returns
+a dictionary containing a custom key/value pair as the third return value in
+``evaluate``.
+
+Scaling federated learning
+--------------------------
+
+As a last step in this tutorial, let’s see how we can use Flower to experiment with a
+large number of clients. In the ``pyproject.toml``, increase the number of SuperNodes to
+1000:
+
+.. code-block:: toml
+
+    [tool.flwr.federations.local-simulation]
+    options.num-supernodes = 1000
+
+Note that we can reuse the ``ClientApp`` for different ``num-supernodes`` since the
+Context is defined by the ``num-partitions`` argument in the ``client_fn()`` and for
+simulations with Flower, the number of partitions is equal to the number of SuperNodes.
+
+We now have 1000 partitions, each holding 45 training and 5 validation examples. Given
+that the number of training examples on each client is quite small, we should probably
+train the model a bit longer, so we configure the clients to perform 3 local training
+epochs. We should also adjust the fraction of clients selected for training during each
+round (we don’t want all 1000 clients participating in every round), so we adjust
+``fraction_fit`` to ``0.025``, which means that only 2.5% of available clients (so 25
+clients) will be selected for training each round. We update the ``fraction-fit`` value
+in the ``pyproject.toml``:
+
+.. code-block:: toml
+
+    [tool.flwr.app.config]
+    fraction-fit = 0.025
+
+Then, we update the ``fit_config`` and ``server_fn`` functions in ``server_app.py`` to
+the following:
+
+.. code-block:: python
+
+    def fit_config(server_round: int):
+        config = {
+            "server_round": server_round,
+            "local_epochs": 3,
+        }
+        return config
+
+
+    def server_fn(context: Context):
+        # Read from config
+        num_rounds = context.run_config["num-server-rounds"]
+        fraction_fit = context.run_config["fraction-fit"]
+
+        # Initialize model parameters
+        ndarrays = get_weights(Net())
+        parameters = ndarrays_to_parameters(ndarrays)
+
+        # Create FedAvg strategy
+        strategy = FedAvg(
+            fraction_fit=fraction_fit,  # Train on 25 clients (each round)
+            fraction_evaluate=0.05,  # Evaluate on 50 clients (each round)
+            min_fit_clients=20,
+            min_evaluate_clients=40,
+            min_available_clients=1000,
```

Comment:
I'd flag this as optional changes to apply. Technically speaking, this isn't strictly needed and might make people think this needs to match the number of supernodes set in `pyproject.toml`.